pkg/operator/machinedeletionhooks/machinedeletionhooks.go

```diff
@@ -0,0 +1,352 @@
+	desiredControlPlaneReplicasCount, err := readDesiredControlPlaneReplicasCount(c.configMapListerForKubeSystemNamespace)
+	if err != nil {
+		return err
+	}
+	masterMachines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	if err != nil {
+		return err
+	}
+	// machines with the deletion hooks are the ones that host etcd members
+	memberMachines := filterMachinesWithMachineDeletionHook(masterMachines)
+
+	// if the number of master machines is > controlPlane.replicas based on data from the cache
+	if len(memberMachines) <= desiredControlPlaneReplicasCount {
+		klog.V(4).Infof("haven't found a replacement machine, the number of desired control plane replicas: %d must be greater than the current number of machines that host an etcd member: %d", desiredControlPlaneReplicasCount, len(memberMachines))
+		return nil
+	}
+
+	// if at least one member machine is pending deletion
+	memberMachinesPendingDeletion := filterMachinesPendingDeletion(memberMachines)
+	if len(memberMachinesPendingDeletion) == 0 {
+		klog.V(4).Infof("currently, we don't have a machine pending deletion")
+		return nil
+	}
+
+	// get the voting members
+	// issue a live call instead of reading from the config map
+	// because we need to filter out learners
+	members, err := c.etcdClient.MemberList(ctx)
+	if err != nil {
+		return err
+	}
+	var currentVotingMemberIPListSet sets.String
+	var currentLearnerIPListSet sets.String
+	for _, member := range members {
+		votingMemberIP, err := ceohelpers.MemberToNodeInternalIP(member)
+		if err != nil {
+			return err
+		}
+		if member.IsLearner {
+			currentLearnerIPListSet.Insert(votingMemberIP)
+			continue
+		}
+		currentVotingMemberIPListSet.Insert(votingMemberIP)
+	}
+
+	// if the number of voting member is equal controlPlane.replicas based on data in the config map
+	// that means all API servers have already removed a member from their configuration,
+	// and we can safely carry on
+	if currentVotingMemberIPListSet.Len() != desiredControlPlaneReplicasCount {
+		klog.V(4).Infof("the number of voting members: %d is not equal to the number of desired control plane replicas: %d", currentVotingMemberIPListSet.Len(), desiredControlPlaneReplicasCount)
+		return nil
+	}
+
+	// find a machine that doesn't have a member
+	for memberMachinePendingDeletionIP, memberMachinePendingDeletion := range indexMachinesByNodeInternalIP(memberMachinesPendingDeletion) {
+		if !currentVotingMemberIPListSet.Has(memberMachinePendingDeletionIP) && !currentLearnerIPListSet.Has(memberMachinePendingDeletionIP) {
+			memberMachinePendingDeletionCopy := memberMachinePendingDeletion.DeepCopy()
+			removeMachineDeletionHook(memberMachinePendingDeletionCopy)
+			if _, err := c.machineClient.Update(ctx, memberMachinePendingDeletionCopy, metav1.UpdateOptions{}); err == nil {
+				klog.Infof("successfully removed the deletion hook from machine %v", memberMachinePendingDeletion.Name)
```

Comment:
do we want to record an event?

Reply:
I think yes it's probably worth noting when we unblock a machine.